Community/Expire/automation/jsr223/python/community/expire/expire.py

```diff
@@ -103,127 +106,239 @@ def get_config(item_name):
-    metadata value set.
-    """
-    def generate_triggers(function):
-        for item_name in [i for i in items if get_value(i, "expire")]:
-            try:
-                get_config(item_name)
-            except AssertionError:
-                init_logger.log.error("Expire config on {} is not valied: {}"
-                                      .format(item_name,
-                                              get_value(item_name, "expire")))
-            else:
-                when("Item {} received update".format(item_name))(function)
-        return function
-    return generate_triggers
 
-@log_traceback
 def expired(item, exp_type, exp_state, log):
     """
-    Called when an Item expires. postUpdate or sendCommand to the configured
+    Called when an Item expires, postUpdate or sendCommand to the configured
     state.
 
     Arguments:
         - item: The Item that expired.
-        - cfg: Contians a dict representation of the expire config returned by
-        get_config.
-        - log: Logger from the expire Rule.
+        - exp_type: The action type, 'state' or 'command'.
+        - exp_state: The state to expire to.
+        - log: Logger from the expire rule.
     """
-    log.debug("{} expired, {} to {}".format(item.name, exp_type, exp_state))
-
-    # Force the state to a StringType for StringItems to allow us to set the
-    # Item to "UNDEF" and "NULL" as opposed to the UnDefTypes.
-    if item.type == "String" and isinstance(exp_state, basestring):
-        exp_state = StringType(exp_state)
+    log.debug("'{}' expired, {} '{}'".format(item.name, exp_type, exp_state))
 
     if exp_type == "state":
         events.postUpdate(item, exp_state)
-    else:
+    elif exp_type == "command":
         events.sendCommand(item, exp_state)
+    else:
+        log.warn("Unrecognized action '{}' for item '{}'"
+                 .format(exp_type, item.name))
+
 
-@rule("Expire",
-      description=("Simulates the Expire1 binding, updating or commanding an "
-                   "Item after a comnfigured amount of time."),
-       tags=["expire"])
-@trigger_generator()
 def expire(event):
     """
-    Called when a member of the configured Group changes. The name of the Group
-    is imported from configuration as expire_items_gr.
+    Called when an item configured for expire receives an update.
 
     If the Item updates to an UnDefType, the change is ignored.
     If the Item updates to the same state configured in the expire config any
     expire timer is cancelled.
     If the Item updates to a different state configured in the expire config a
     new timer is created to go off at the configured time in the future.
     """
+    item_name = event.itemName
+
     # Ignore changes to an UnDefType.
     if isinstance(event.itemState, UnDefType):
+        if (timers[item_name] is not None
+                and not timers[item_name].hasTerminated()):
+            timers[item_name].cancel()
```

Comment:
This does change the behavior somewhat from the Expire binding IIRC. As the docstring indicates, UnDefType changes are in fact ignored (i.e. existing Timers are left running). I agree that this behavior is probably more desirable, but the docstring needs to be updated. The events are not just ignored, any open timers are now cancelled.

Reply:
Interesting, I was not aware of that. I'm just on my phone now but I think this used to set the timer to `None`.

Reply:
The original used to just return and leave the Timer unchanged. But I think canceling the Timer is a better way to handle this. We've already deviated from Expire a bit with the String Item, I see no reason why we shouldn't change this too. But the docstring is now incorrect.

Reply:
I agree, that behaviour of the expire binding seems strange. I did not go over this with a fine toothed comb before submitting as I didn't have time. I wanted to submit it before we left for the weekend so that you would be able to see it and test and offer comment.

Reply:
So I'm making the changes we discussed and I've come to this one...
Do we cancel the timer if the item changes to `UnDefType` or do we treat `UnDefType` as a state like any other and just remove this check altogether?

Reply:
Let's do what I think we both think makes sense and have UnDefType cancel the timer. It just doesn't make sense to me that you would use Expire to change an UNDEF or NULL to some other state. And either way there will be sue cases we can't support.

Reply:
Sounds good to me. I'll go over the docstrings tonight and push the changes.